src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,91 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int INITIAL_CAPACITY = 10;
+    private static final double INCREMENT_CAPACITY = 1.5;
+    private int size;
+    private T[] elements;
+
+    public ArrayList() {
+        elements = (T[]) new Object[INITIAL_CAPACITY];
+    }
+
+    private void grow() {
+        T[] tempArray = (T[]) new Object[(int) (elements.length * INCREMENT_CAPACITY)];
+        System.arraycopy(elements, 0, tempArray, 0, size);
+        elements = tempArray;
+    }
+
+    private void checkArrayBoundary(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index is invalid");
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkArrayBoundary(index);
+        if (size == elements.length) {
+            grow();
```

Comment:
we could move this check if `(size == elements.length)` into `grow` method(this will reduce number if duplicated code). If we do this, then it might be worth to rename `grow` method to `ensureCapacity` or smth like this

Reply:
I did it and named method 'checkFillingArray'